Add Nginx Reverse Proxy guide for Pi-hole v6 (#1379)
pi-hole/docs: docs/guides/reverse_proxy/nginx.md

<!-- markdownlint-disable code-block-style -->
!!! info "Key Changes in v6"
    Unlike Pi-hole v5, version 6 no longer relies on `lighttpd`. The new architecture features a built-in web server within `pihole-FTL`. This change significantly simplifies reverse proxy configurations as there are no longer conflicting external web server dependencies to manage by default.

    Since Pi-hole v6 handles its own web interface and API directly, `Nginx` simply acts as a (secure) gateway, without the need for complex fastcgi or PHP-fpm passthroughs.
<!-- markdownlint-enable code-block-style -->

<!-- markdownlint-disable code-block-style -->
!!! danger "Security Warning: Exposure to the Internet"
    Exposing your Pi-hole directly to the open internet is **strongly discouraged**. Doing so can make your instance a target for unauthorized access and cyber-attacks. If you choose to do so, it is your responsibility to follow security best practices, including for example, but not limited to:

    - Using strong, unique passwords;
    - Implementing Multi-Factor Authentication (MFA);
    - Restricting access via VPN or IP whitelisting;
    - etc

>     **Disclaimer:** Utilizing `Nginx` as a reverse proxy does not automatically guarantee safety against all vulnerabilities associated with internet exposure. This guide focuses exclusively on the generic configuration of `Nginx` as a web proxy with some security tips; it does not cover, nor is it intended to cover, the advanced hardening techniques required for public internet exposure. Neither this guide nor the Pi-hole team shall be held responsible for any security breaches, data loss, or damages resulting from your network configuration.

Reviewer: This liability disclaimer sounds a bit to strong for me, I'd like to get the opinion of someone else @pi-hole/docs-maintainers.

Reviewer 2: I think we should in general not suggest that it is a good idea to expose Pi-hole to the internet. No matter if behind reverse proxy or not.

Author: Too Strong? .... I placed it especially to cover any possible issue and to underline how exposing the service to internet is a thing that should be done _ONLY_ under the right knowledge... but well, on this point, you guys have _green light_ for me whatever you decide 😊 edit as you wish

Reviewer: I'm not sure if this phrasing would be good, this is why I checked in

Reviewer: > Too Strong? .... I placed it especially to cover any possible issue and to underline how exposing the service to internet is a thing that should be done _ONLY_ under the right knowledge... but well, on this point, you guys have _green light_ for me whatever you decide 😊 edit as you wish

I was more referring to the "absolutely no liability" part, of which I was unsure about.
The previous one about the dangers is reasonable.

Author: > I was more referring to the "absolutely no liability" part, of which I was unsure about.
> The previous one about the dangers is reasonable.

Belong to me, it isnt in any case ("by design") team's accountability for anything done or consequences produced by the software improperly used. This phrase is just merely "repeating"/"underling" something really objectives.
If "my" (AI-revised ahah) english wasnt good, I tried to just discourage people to expose it on internet and (if they do), this guide doesnt cover this scenario - it is up to them; this was my target.
Sorry in case Im misunderstanding you.

Reviewer: I was comparing it to the warning at https://docs.pi-hole.net/ftldns/interfaces/. Neither exposing port 53 nor 80/443/whatever HTTP(S) is recommended. In my eyes "Neither this guide nor the Pi-hole team shall be held responsible for any security breaches, data loss, or damages resulting from your network configuration." sounds a bit overly defensive, but I am not quite sure on how to proceed. That's why I wanted to get another opinion here.

    `Nginx` requires a restart or reload to apply updated certificates, a process that is not handled automatically. Managing this would require cumbersome scripts and timers; therefore, it is highly recommended to generate certificates independently using dedicated tools like **Certbot** or similar, which can use hooks to reload `Nginx` seamlessly upon renewal.
<!-- markdownlint-enable code-block-style -->

<!-- markdownlint-disable code-block-style -->
!!! security "Security Note"
    Since `Nginx` will handle SSL/TLS termination and encryption for external traffic, we will not connect to Pi-hole's native SSL/TLS for the internal connection. This simplifies the communication between `Nginx` and Pi-hole (localhost), as encryption is not strictly required for traffic that never leaves the machine.

    By binding the Pi-hole dashboard to `127.0.0.1` only, it is no longer exposed to your entire network. Only `Nginx` (running on the same machine) can "talk" to it. This effectively creates a "security air-lock."
<!-- markdownlint-enable code-block-style -->

### Update Pi-hole's Configs

> You must instruct Pi-hole to stop listening on port 80, 443 or any other ports and switch to our (arbitrary choosen) port **8053**. Additionally, for sake of security, we will restrict its web interface to listen only on the local interface (`127.0.0.1`), ensuring it is only accessible via the `Nginx` proxy.

Reviewer: ```suggestion
You must instruct Pi-hole to stop listening on port 80, 443 or any other ports and switch to our (arbitrarily chosen) port **8053**. Additionally, for sake of security, we will restrict its web interface to listen only on the local interface (`127.0.0.1`), ensuring it is only accessible via the `Nginx` proxy.
```

1. **Install Nginx**

    ```bash
    sudo apt update
    sudo apt install -y nginx nginx-extras
    ```

2. **Clean up default configurations and restart**

    Before proceeding adding the Pi-hole dedicated configurations, it is recommended to remove or at least deactivate, according to your distribution configuration, the default *"Welcome to Nginx"* site to avoid potential/hypothetical conflicts.

    ```bash
    sudo systemctl restart nginx.service
    ```

### Configure the HTTP redirect to HTTPS (Optional)

This configuration is intended **only** for users who wish to automatically redirect incoming traffic from a non-encrypted port to the secure HTTPS port.

In this example, we will use port **8080** as the external HTTP to avoid possible conflicts with other services; it will "force" the redirect of all traffic toward the secure port **453** (used in this guide).

!!! info "Scope"
    This redirect applies only to the **external/exposed** ports. Internal communication between `Nginx` and Pi-hole remains on port `8053` and is unaffected by this rule.

1. **Create the redirect file configuration**

>     Place the following file in your Nginx configuration path (e.g., `/etc/nginx/conf.d/` or `/etc/nginx/sites-available/` depending on your distribution configuration).

Reviewer: maybe https://github.com/pi-hole/docs/pull/1379/changes#r3248200476 can go here

Author: Yeah, I copy/pasted to be consistent in the wording, but is the same thought of before 🤔

        server_name _; # Or your domain/IP

        # Redirect all HTTP requests to HTTPS
        return 301 https://$host:453$request_uri;
    }
    ```

3. **Enable, test and restart `Nginx`**

    Once the `pihole_v6_http_redirect.conf` file is saved and enabled, verify configurations and restart the service:

    ```bash
    sudo nginx -t
    sudo systemctl restart nginx.service
    ```

!!! tip "Firewall Reminder"
    If you are using a firewall (like `ufw`), be sure to allow traffic on port **8080** or the redirection won't work:
    `sudo ufw allow 8080/tcp`

<!-- markdownlint-disable code-block-style -->
!!! warning "Connection Refused / Unable to Connect"
    At this stage of the guide, if your browser successfully redirects you to the HTTPS port but then shows **"Unable to Connect"**, don't panic. This occurs because the `Nginx` HTTP redirection is active, but the HTTPS Proxy and the relative bridge to the backend isn't configured yet.

    You must now proceed to the next step: [Configure the HTTPS Proxy ](#configure-the-https-proxy).
<!-- markdownlint-enable code-block-style -->

### Configure the HTTPS Proxy

1. **Generate the required Pi-hole SSL/TLS files**

>     We generate the private key and a Certificate Signing Request (CSR). In real world, we should send the generated CSR to the CA for the signing, getting back the Certificate; in this guide (at command #4) we directly self-sign them, or in a Private PKI environments, we autonomusly sign it with the Private CA.

Reviewer: ```suggestion
    We generate the private key and a Certificate Signing Request (CSR). In the real world, we should send the generated CSR to the CA for the signing, getting back the Certificate; in this guide (at command #4) we directly self-sign them, or in a Private PKI environments, we autonomously sign it with the Private CA.
```

    sudo openssl req -sha512 -nodes -new \
      -subj "/C=IT/ST=Italy/L=Milan/O=Pi-hole/CN=pi.hole" \
      -addext "subjectAltName = DNS:pi.hole,DNS:*.pi.hole" \
      -key /etc/nginx/ssl/pihole.key \
      -out /etc/nginx/ssl/pihole.csr
    # Read Pi-hole CSR
    sudo openssl req -text -noout -in /etc/nginx/ssl/pihole.csr

    # 4. Self-sign the Pi-hole certificate (Valid for 1 years)
    echo "subjectKeyIdentifier=hash" > /tmp/openssl.cnf
    echo "authorityKeyIdentifier=keyid:always,issuer" >> /tmp/openssl.cnf
    echo "basicConstraints=critical,CA:FALSE" >> /tmp/openssl.cnf
    echo "keyUsage=critical,digitalSignature,keyEncipherment" >> /tmp/openssl.cnf
    echo "extendedKeyUsage=serverAuth" >> /tmp/openssl.cnf
    echo "subjectAltName=DNS:pi.hole,DNS:*.pi.hole" >> /tmp/openssl.cnf
    sudo openssl x509 -req -sha512 -days 365 \
      -signkey /etc/nginx/ssl/pihole.key \
      -in /etc/nginx/ssl/pihole.csr \
      -out /etc/nginx/ssl/pihole.crt \
      -extfile /tmp/openssl.cnf
    sudo rm -rf /tmp/openssl.cnf
    # Read Pi-hole certificate
    sudo openssl x509 -text -noout -in /etc/nginx/ssl/pihole.crt
    ```

2. **Create the proxy file configuration**

>     Place the following file in your `Nginx` configuration path (e.g., `/etc/nginx/conf.d/` or `/etc/nginx/sites-available/` depending on your distribution configuration).

Reviewer: I think it'd be useful to tell the user to look for potential port conflicts in `server` blocks already existing in `/etc/nginx/conf`, `/etc/nginx/conf.d/` and `/etc/nginx/sites-available/`. Both with Pi-hole and other running services.

Author: For this topic, I have tried to forecast the possibility and adopted the port `8080` in the config, to prevent possible conflicts for the port `80`.
To be honest, I think that this should go a bit under each person custom scenarios. Especially for that I wrote in the `Network Mapping & Port Alignment` section which port the guide would use, and place attention to it, change according to personal setup.
At the start I wrote the guide as copy/paste, but thanks to your feedback, I rework it as "guidance"; as we said before in the previous interactions, if a person doesn't know what Nginx is or how to configure it, it is way better to use standard Pi-hole feature. This is a "guidance" for people that have knowledge of Nginx and are looking for a "pre-made" base for the Reverse Proxy.
Pi-hole v5 Nginx reverse proxy was a bit more complex due to `fastcgi` and `php`... but know it is way easier; I have find out many post, but most of the time, same questions: redirect api? redirect only admin? and so on. So I think we may do not focus extremely too much, because if they skip (important) section of the guide, we can't really repeat 10 times all the sections again... this is my personal point of view.

In case of a fresh install, it is up to the user to configure it wisely. Otherwise it is just used for.... ? In the "marketing phrase" we have removed I tried to underlined how Nginx is just not to be used as "Im cool guy, I have a reverse proxy", cause it adds nothing more than what Pi-hole (v6 especially) already provides. The team did an amazing job improving the internal http web server. Im to give support to test "new technologies", but not so much to "copy/paste" and forget... I wrote sections about possible issues in case of others pre-existing directives, using also "default Nginx value", just because lower value (in custom environment) may break and requires troubleshooting (as I had to do - in a customized nginx environment). So Im not really sure about this point 🤔 (just because I spammed already of warning everywhere in each dedicated sections 😂 it is becoming a warning spam, more than guide for how much I tried to underline to pay attention ahahah)

Reviewer: Thinking more about "To avoid conflicts with Nginx or other services running on your system, we need to move Pi-hole v6 away from the standard HTTP port (80).", I guess its fine like this.

Reviewer 2: Maybe a native speaker @PromoFaux can give an opinion?

        # --- HTTP/2 SUPPORT ---
        # For Nginx >= 1.25.1: Use the standalone directive below
        http2 on; 
        # For Nginx < 1.25.1: Comment "http2 on" above and 
        # append "http2" to the listen lines (e.g., listen 453 ssl http2;)
        
        server_name _; # Or your domain/IP

        # --- SSL/TLS ---
        ssl_certificate_key /etc/nginx/ssl/pihole.key;
        ssl_certificate /etc/nginx/ssl/pihole.crt;
        # Intermediate Mozilla's security standards
        ssl_protocols TLSv1.2 TLSv1.3;
        ssl_ecdh_curve X25519MLKEM768:X25519:prime256v1:secp384r1:secp521r1;
        ssl_ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305;
        ssl_prefer_server_ciphers off;

        # --- SECURITY ---
        # Restricted HTTP Methods: Allow only methods mentioned in API docs
        # and block everything else (e.g., TRACE, CONNECT);
        # this prevents potential exploit attempts using uncommon request methods.
        if ($request_method !~ ^(GET|HEAD|POST|PATCH|PUT|DELETE)$ ) { return 405; }

>         # HSTS (ngx_http_headers_module is required) (63072000 seconds)

Reviewer: ```suggestion
        # HSTS (ngx_http_headers_module is required) is remembered for 63072000 seconds (2 years)
```

Even though it's mentioned the the Security headers below, the comment is a good place to clarify how long 63072000s actually are.

>         add_header Strict-Transport-Security "max-age=63072000; includeSubDomains" always;

Reviewer: ```suggestion
        add_header Strict-Transport-Security "max-age=63072000" always;
```

I don't think we should affect other subdomains the user might have and potentially lock them out of their http sites. (If they want to use https://pihole.mydomain.net and then http://other.mydomain.net)

Author: Uhm... about it... I would recommend to use `includeSubDomains` (HTTPS always better to HTTP), but to not be too strict and leave opening in people's configurations, it is ok for me as well.

This would require a rephrasing (or simply the removal) of the phrase:
```
The `includeSubDomains` flag extends this protection to all related subdomains.
```

How may I do it? - Not really used to all "feature" github provides for reviewing/add comments

Reviewer: ```suggestion
        # Remove "; includeSubDomains" if any of your subdomains are using plain HTTP (e.g. if you use https://pihole.mydomain.net and http://other.mydomain.net)
        add_header Strict-Transport-Security "max-age=63072000; includeSubDomains" always;
```
I guess a comment would also be fine. It is already mentioned in the "Security Headers" section.

@DoctorD90 the suggestion feature in GitHub is sometimes a bit finicky, to use it select all lines you  wish to change beforehand and then write only the new contents between `` ```suggestion `` and `` ``` ``. There is also a [+-] icon in the bar for pre-filling the suggestion code block:

<img width="663" height="302" alt="Image" src="https://github.com/user-attachments/assets/5af61c1d-20cf-402e-a63f-4dd242578554" />

Suggestions are mostly only useful for small changes that can be applied directly from GitHub web.

Author: > Suggestions are mostly only useful for small changes that can be applied directly from GitHub web.

THANK YOU VERY MUCH! Im almost used to command line, so Im new to all the web feature.

> I don't think we should affect other subdomains the user might have and potentially lock them out of their http sites.

If you agree, I would go with your first suggestion; as I said before, I agree with you. This is in in the end a Nginx Reverse Proxy guide dedicated to Pi-hole, not a general purpose hardening one :) In the case you agree, I will remove also the line in the Security Headers description.

!!! note "Version Check"
    You can check your currently installed version by running `sudo nginx -v`.

##### SSL/TLS Certificates

In this guide, we have created a *Self Signed Certificate* as explained the process. You are invited to use your own certificate, self-signed or signed by a Public/Private CA, customize the configuration accordingly.

Furthermore, this configuration follows the [intermediate Mozilla's security standards](https://ssl-config.mozilla.org/#server=nginx&version=1.27.3&config=intermediate&openssl=4.0.0&hsts&guideline=6.0). You can review the settings or stay updated by visiting the [Mozilla SSL Configuration Generator](https://ssl-config.mozilla.org).

##### HTTP Methods Filtering

To harden your setup, we implement a filter on the types of HTTP requests the server will accept; the `$request_method` line acts as a security gatekeeper; this minimizes the attack surface by blocking unexpected request types.

The `$request_method` filtering in the configuration above is strictly aligned with the [HTTP methods reported by the Pi-hole API documentation](/api/); any other method will return a `405 Method Not Allowed` error.

##### Security Headers

> These header can add a layer of protection by instructing the user's browser how to handle the site's content securely:

Reviewer: ```suggestion
These headers can add a layer of protection by instructing the user's browser how to handle the site's content securely:
```

> - **Strict-Transport-Security:** Forces the browser to communicate with the server exclusively over HTTPS. By setting a `max-age` (in this case, 2 years), it ensures that even if you try to access the site via HTTP, the browser will automatically redirect to the secure version before any data is sent. The `includeSubDomains` flag extends this protection to all related subdomains.

Author: ```suggestion
- **Strict-Transport-Security:** Forces the browser to communicate with the server exclusively over HTTPS. By setting a `max-age` (in this case, 2 years), it ensures that even if you try to access the site via HTTP, the browser will automatically redirect to the secure version before any data is sent.
```
Related to https://github.com/pi-hole/docs/pull/1379/changes#r3248159474

    ```bash
    sudo mkdir -p /etc/nginx/ssl
    ```

2. **Create a Minimal Private CA Certificate (Test Purpose Only)**
    Even for a single user, TLS protocol requires the client certificate to be signed by a CA. Here we generate a self-signed Private CA using the **ECDSA** (Elliptic Curve) algorithm.

    ```bash
    # Generate CA private key (using NIST P-521 curve)
    sudo openssl ecparam -genkey -name secp521r1 -out /etc/nginx/ssl/CertificateAuthority.key
    # Read CA private key
    sudo openssl pkey -text -noout -in /etc/nginx/ssl/CertificateAuthority.key

    # Generate the CA certificate (Self-signed, valid for 10 years)
    sudo openssl req -x509 -sha512 -nodes -days 3650 \
      -key /etc/nginx/ssl/CertificateAuthority.key \
      -out /etc/nginx/ssl/CertificateAuthority.pem \
      -subj "/C=IT/ST=Italy/L=Milan/O=MyOrg/CN=MyPrivateCA" \
      -addext "subjectKeyIdentifier=hash" \
      -addext "authorityKeyIdentifier=keyid:always,issuer" \
      -addext "basicConstraints=critical,CA:true"
    # Read CA certificate
    sudo openssl x509 -text -noout -in /etc/nginx/ssl/CertificateAuthority.pem
    ```

3. **Generate the Client Certificate**

>     We generate the private key for the user and a Certificate Signing Request (CSR). In real world, we should send the generated CSR to the CA for the signing, getting back the Certificate; in this guide (at command #3), and in Private PKI environments, we autonomusly sign it with our Private CA.

Reviewer: ```suggestion
    We generate the private key for the user and a Certificate Signing Request (CSR). In the real world, we should send the generated CSR to the CA for the signing, getting back the Certificate; in this guide (at command #3), and in Private PKI environments, we autonomously sign it with our Private CA.
```

    # 3. Sign the Client certificate with our Private CA (Valid for 1 years)
    sudo openssl x509 -req -sha512 -days 365 \
      -CA /etc/nginx/ssl/CertificateAuthority.pem \
      -CAkey /etc/nginx/ssl/CertificateAuthority.key \
      -in client.csr -out client.crt
    # Adjust ownership of the generated certificate
    sudo chown $(id -u):$(id -g) client.crt
    # Read Client certificate
    openssl x509 -text -noout -in client.crt
    ```

4. **Export to PKCS#12 for Browser Installation**
    To use this certificate in a browser (Chrome, Firefox, Safari), must be converted into a `.p12` bundle.

    ```bash
    # Bundle Client data to PKCS#12
    openssl pkcs12 -export \
      -certfile /etc/nginx/ssl/CertificateAuthority.pem \
      -inkey client.key -in client.crt \
      -out client.p12 \
      -macalg SHA512 -iter 100000 \
      -keypbe AES-256-CBC -certpbe AES-256-CBC \
      -pbmac1_pbkdf2 -pbmac1_pbkdf2_md SHA512 \
      -name "MyPiholeLoginCertificate" \
      -passout pass:abcdef123456
    # Read Client p12 certificate
    openssl pkcs12 -info -in client.p12 \
      -nodes -passin pass:abcdef123456
    ```

    !!! danger "Change the PKCS#12 password"

>         In this guide, to speed up the command execution, the **weak** password `abcdef123456` has been used; we reccomend to choose a strong password and do not pass it via command line to keep it as secret as possible.

Reviewer: ```suggestion
        In this guide, to speed up the command execution, the **weak** password `abcdef123456` has been used; we recommend to choose a strong password and do not pass it via command line to keep it as secret as possible.
```

    !!! tip "TLS/SSL Installation"
        - **Trust:** You may also need/want to import the `CertificateAuthority.pem` as a Trusted Root CA if it is a private one. Detailed instructions can be found in the [Pi-hole Official Documentation](/api/tls/#adding-the-ca-to-your-browser);

>         - **Browser:** Import the `client.p12` file into your browser's certificate manager (Settings > Security > Manage Certificates > Your Certificates); it is similar to precedent point, so you can read the [Pi-hole Official Documentation](/api/tls/#adding-the-ca-to-your-browser);

Reviewer: ```suggestion
        - **Browser:** Import the `client.p12` file into your browser's certificate manager (Settings > Security > Manage Certificates > Your Certificates); it is similar to the previous point, so you can read the [Pi-hole Official Documentation](/api/tls/#adding-the-ca-to-your-browser);
```
